scipy/signal/_signaltools.py

```diff
@@ -4625,3 +4625,128 @@ def decimate(x, q, n=None, ftype='iir', axis=-1, zero_phase=True):
         sl[axis] = slice(None, None, q)
 
     return y[tuple(sl)]
+
+def envelope(x, N=None, method='analytic'):
+    """
+    Calculate the upper and lower envelopes of a 1-D signal.
+
+    By default, the envelopes of `x` are calculated by the magnitude of its
+    analytical signal.
+
+    Parameters
+    ----------
+    x : array_like
+        The 1-D array holding the signal data for which the envelopes are
+        calculated.
+    N : int
+        If `method` is `'analytic'`, `N` represents the number of Fourier components.
+        If `method` is `'rms'`, `N` is the sliding window length.
+        If `method` is `'peak'`, `N` is the required minimal horizontal distance
+        between neighboring peaks. For `'rms'` and `'peak'` methods, `N` must be
+        specified and must be positive. Default: ``x.shape[0]``
```

Comment:
I feel like this parameter might be doing a little bit too much heavy lifting API wise. I think a common pattern, e.g. used in `numpy.pad` is to split these additional optional parameters that depend on the correct mode. E.g. in this case:
```python
def envelope(x, *, method='analytic', rms_length=None, peak_distance=None):
    ...
```

Alternatively, what do you think about splitting this function into three: `envelope_rms`, `envelope_analytic`, `envelope_peak`? Looking at the docstring and code, there wouldn't be too much duplication and this would make each function easier to reason about and document.